pkgs/top-level/all-packages.nix

```diff
@@ -934,6 +934,8 @@ with pkgs;
 
   dotnetCorePackages = recurseIntoAttrs (callPackage ../development/compilers/dotnet {});
 
+  dotnet8Packages = recurseIntoAttrs (callPackage ../development/compilers/dotnet/sdk/8 {});
```

Comment:
I'd prefer a more generic name. Or maybe keep the source build sdk in dotnetCorePackages, but with a `srcBuild` prefix, ie. `dotnetCorePackagaes.srcBuild.sdk_8_0`

Reply:
I was going for something along the lines of `python3Packages`, but I'm open to alternatives.

I wanted to combine the sdk/runtime/aspnetcore/etc in one place for a given version and (source vs binary).

I could put it all under `dotnetCorePackages.srcBuild`, but:

- I'm not sure if 'dotnetCore' was mean to refer to .NET Core the product, or simply be the core packages for the dotnet ecosystem.  If the former, I think the name is obsolete and it's generally referred to as just 'dotnet' now.  I imagined we'd be able to eventually merge/replace the old `dotnetPackages` with `dotnet8Packages`, and maintain multiple major versions in parallel.  If it's the latter, then I don't mind using the same name, but I don't see anything else in nixpkgs using `...CorePackages`.

- I think the source build should take precedence, e.g. `firefox` vs. `firefox-bin`.

I'm aiming for something where you can do (similar to python) `dotnet8Packages.buildDotnetModule {}` without specifying sdk/runtime, and having the source and binary versions compatible so you could do:  `{ dotnet8Packages = dotnet8Packages-bin; }

I thought it was best to avoid doing anything to the existing packages in this change.

Reply:
Oh, that's a good idea actually. I think the Core referred to the .NET Core product, to avoid conflicts with the dotnetPackages namespace.

I think we can leave the dotnet8Packages then. Are the binary 8.0 build also available under this name?

And also I wonder what should we do in case we'll want to have multiple minor versions at once (like 8.0 and 8.1), but I don't think something like that will be needed often.

Overall I agree with you, and the whole dotnet package naming needs a big refactor, unfortunately that will be hard to do without breaking changes :)

Reply:
> Are the binary 8.0 build also available under this name?

No, the new attribute contains only the source built packages.  The way I imagined that working is that the binary ones would be under something like `dotnet8Packages-bin` (I don't love that naming, but it has a precedent).

> And also I wonder what should we do in case we'll want to have multiple minor versions at once (like 8.0 and 8.1), but I don't think something like that will be needed often.

Python does `python311` / `python311Packages`.  I don't really like that because the versions are ambiguous.

Perhaps we should go with `dotnetPackages_8`, and then `dotnetPackages_8_0` if we ever need minor versions?

There are actually quite a few things like that already: `chickenPackages_5`, `rubyPackages_3_1`, `llvmPackages_14`.

We currently do:

```
  dotnet-sdk_6 = dotnetCorePackages.sdk_6_0;
  dotnet-sdk_7 = dotnetCorePackages.sdk_7_0;
  dotnet-sdk_8 = dotnetCorePackages.sdk_8_0;

  dotnet-runtime_6 = dotnetCorePackages.runtime_6_0;
  dotnet-runtime_7 = dotnetCorePackages.runtime_7_0;
  dotnet-runtime_8 = dotnetCorePackages.runtime_8_0;

  dotnet-aspnetcore_6 = dotnetCorePackages.aspnetcore_6_0;
  dotnet-aspnetcore_7 = dotnetCorePackages.aspnetcore_7_0;
  dotnet-aspnetcore_8 = dotnetCorePackages.aspnetcore_8_0;

  dotnet-sdk = dotnetCorePackages.sdk_6_0;
  dotnet-runtime = dotnetCorePackages.runtime_6_0;
  dotnet-aspnetcore = dotnetCorePackages.aspnetcore_6_0;
```

And that could become:

```
  dotnet-sdk_6 = dotnetPackages_6.sdk
  dotnet-sdk_7 = dotnetPackages_7.sdk
  dotnet-sdk_8 = dotnetPackages_8.sdk

  dotnet-runtime_6 = dotnetPackages_6.runtime
  dotnet-runtime_7 = dotnetPackages_7.runtime
  dotnet-runtime_8 = dotnetPackages_8.runtime

  dotnet-aspnetcore_6 = dotnetPackages_6.aspnetcore
  dotnet-aspnetcore_7 = dotnetPackages_7.aspnetcore
  dotnet-aspnetcore_8 = dotnetPackages_8.aspnetcore

  dotnet-sdk = dotnetPackages_6.sdk
  dotnet-runtime = dotnetPackages_6.runtime
  dotnet-aspnetcore = dotnetPackages_6.aspnetcore
```

Reply:
I guess we can go with `dotnetPackages_8`. Not a fan of the name, but it solves the ambiguity problem.

I think the .NET 8.0 binary versions should be also available there, ie. `dotnetPackages_8.sdk-bin`. Then, making the switch over should be a little bit easier.

Reply:
> I guess we can go with dotnetPackages_8. Not a fan of the name, but it solves the ambiguity problem.

Do you mean you prefer `dotnet8Packages`?  I think I prefer the suffix as it seems to be more commonly done that way, and I wouldn't want to add a minor version without a separator (like python).

> I think the .NET 8.0 binary versions should be also available there, ie. dotnetPackages_8.sdk-bin. Then, making the switch over should be a little bit easier.

Before switching any packages over I think we should change the binary packages to also be structured in a similar way to the packages here.  In which case, we shouldn't really have to reference the sdk directly very often.

So imagine we had:

```
dotnetPackages_8-bin = { 
  sdk = dotnetCorePackages.sdk_8_0;
  runtime = dotnetCorePackages.runtime_8_0;
  aspnetcore = dotnetCorePackages.aspnetcore_8_0;
}
```

(This is an example shim.  I doubt we'd actually do it that way, but just to illustrate the new structure...)

A package that might currently do (borrowed from `ryujinx`:

```
{ buildDotnetModule
, dotnetCorePackages
}: buildDotnetModule rec {
  dotnet-sdk = dotnetCorePackages.sdk_8_0;
  dotnet-runtime = dotnetCorePackages.runtime_8_0;
  [...]
```

Could do:

```
{ dotnetPackages_8
}: dotnetPackages_8.buildDotnetModule rec {
```

Then say you want to use the binary version on darwin, you could do:

```
  callPackage ./foo { dotnetPackages_8 = if stdenv.isDarwin then dotnetPackages_8-bin else dotnetPackages_8 };
```